transformer_engine/pytorch/attention.py

```diff
@@ -515,22 +516,42 @@ class AttnFuncWithCP(torch.autograd.Function):
 
     @staticmethod
     def forward(ctx, is_training, q, k, v, cu_seqlens_q, cu_seqlens_k, max_seqlen_q, max_seqlen_k,
-                dropout_p, cp_group, cp_global_ranks, cp_stream, softmax_scale, attn_mask_type,
-                deterministic, use_fused_attention):
+                dropout_p, cp_group, cp_global_ranks, cp_stream, softmax_scale, qkv_format,
+                attn_mask_type, attn_bias_type, attn_bias, deterministic, use_fused_attention):
         if softmax_scale is None:
             softmax_scale = q.shape[-1] ** (-0.5)
 
         cp_size = get_distributed_world_size(cp_group)
         rank = get_distributed_rank(cp_group)
         send_dst = cp_global_ranks[(rank + 1) % cp_size]
-        recv_src = cp_global_ranks[(rank + cp_size - 1) % cp_size]
+        recv_src = cp_global_ranks[(rank - 1) % cp_size]
         batch_p2p_comm = int(os.getenv("NVTE_BATCH_MHA_P2P_COMM", "0")) or (cp_size == 2)
 
         causal = (attn_mask_type == "causal")
```

Comment:
Should this be `causal = ("causal" in attn_mask_type)`? How do we handle the `padding_causal` case?

Reply:
padding cannot be handled yet. In `class FlashAttention` and `class FusedAttention`, I asserted that padding is not supported with CP.

In `FlashAttention`, padding is packed to thd format, so I think after [this PR](https://github.com/NVIDIA/TransformerEngine/pull/641) is merged, it would work.

As for `FusedAttention`, we probably can do same thing as `FlashAttention`, but needs more thinking.